clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,151 @@
+//===--- UseStdMinMaxCheck.cpp - clang-tidy -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "UseStdMinMaxCheck.h"
+#include "../utils/ASTUtils.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/Preprocessor.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+static const llvm::StringRef AlgorithmHeader("<algorithm>");
+
+static bool MinCondition(const BinaryOperator::Opcode &Op, const Expr *CondLhs,
+                         const Expr *CondRhs, const Expr *AssignLhs,
+                         const Expr *AssignRhs, const ASTContext &Context) {
+  if ((Op == BO_LT || Op == BO_LE) &&
+      (tidy::utils::areStatementsIdentical(CondLhs, AssignRhs, Context) &&
+       tidy::utils::areStatementsIdentical(CondRhs, AssignLhs, Context)))
+    return true;
+
+  if ((Op == BO_GT || Op == BO_GE) &&
+      (tidy::utils::areStatementsIdentical(CondLhs, AssignLhs, Context) &&
+       tidy::utils::areStatementsIdentical(CondRhs, AssignRhs, Context)))
+    return true;
+
+  return false;
+}
+
+static bool MaxCondition(const BinaryOperator::Opcode &Op, const Expr *CondLhs,
+                         const Expr *CondRhs, const Expr *AssignLhs,
+                         const Expr *AssignRhs, const ASTContext &Context) {
+  if ((Op == BO_LT || Op == BO_LE) &&
+      (tidy::utils::areStatementsIdentical(CondLhs, AssignLhs, Context) &&
+       tidy::utils::areStatementsIdentical(CondRhs, AssignRhs, Context)))
+    return true;
+
+  if ((Op == BO_GT || Op == BO_GE) &&
+      (tidy::utils::areStatementsIdentical(CondLhs, AssignRhs, Context) &&
+       tidy::utils::areStatementsIdentical(CondRhs, AssignLhs, Context)))
+    return true;
+
+  return false;
+}
+
+static std::string
+CreateReplacement(const bool UseMax, const BinaryOperator::Opcode &Op,
+                  const Expr *CondLhs, const Expr *CondRhs,
+                  const Expr *AssignLhs, const ASTContext &Context,
+                  const SourceManager &Source, const LangOptions &LO,
+                  const StringRef &FunctionName) {
+  const auto CondLhsStr = Lexer::getSourceText(
+      Source.getExpansionRange(CondLhs->getSourceRange()), Source, LO);
+  const auto CondRhsStr = Lexer::getSourceText(
+      Source.getExpansionRange(CondRhs->getSourceRange()), Source, LO);
+  const auto AssignLhsStr = Lexer::getSourceText(
+      Source.getExpansionRange(AssignLhs->getSourceRange()), Source, LO);
+  return (AssignLhsStr + " = " + FunctionName +
+          ((CondLhs->getType() != CondRhs->getType())
+               ? "<" + AssignLhs->getType().getAsString() + ">("
```

Comment:
There is one problem here that is visible in llvm.
Lets take code like this:
```
unsigned char C = 10;
unsigned int B = 257'

if (B < C)
  C = B;
```

this would be changed into `C = std::min<unsigned char>(C, B);`  but thats not valid, as B casted to unsigned char would be smaller than C but not if C is casted to unsigned int and then compared.
Simply as an template argument to std::min we should use wider type, and if we don't fine any, we should read type from already existed implicit cast.

Reply:
@PiotrZSL Can you please guide me for this as I am not sure how to move forward. I am not very good at handling these Implicit casts.